NIFI-840: Create ListS3 processor

adamonduty opened the pull request:
This processor implements an object listing capability for S3. I tried to model it after ListHDFS, but I wasn't able to test the parts that persist state, so I don't know if it works correctly when the primary node in a cluster changes.

jskora commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/ListS3.java:
```diff
@@ -0,0 +1,230 @@
+            currentTimestamp = 0L;
+            currentKeys = new HashSet<>();
+        } else {
+            currentTimestamp = Long.parseLong(stateMap.get(CURRENT_TIMESTAMP));
```
The ListS3 processor throws exceptions on restart if state was cleared through the UI while the processor was stopped.

2016-02-22 14:00:05,586 ERROR [Timer-Driven Process Thread-3] org.apache.nifi.processors.aws.s3.ListS3 ListS3[id=5763acff-1b63-461a-bb72-fd12bd7ac603] ListS3[id=5763acff-1b63-461a-bb72-fd12bd7ac603] failed to process due to java.lang.NumberFormatException: null; rolling back session: java.lang.NumberFormatException: null
2016-02-22 14:00:05,637 ERROR [Timer-Driven Process Thread-3] org.apache.nifi.processors.aws.s3.ListS3 
java.lang.NumberFormatException: null
	at java.lang.Long.parseLong(Long.java:404) ~[na:1.7.0_80]
	at java.lang.Long.parseLong(Long.java:483) ~[na:1.7.0_80]
	at org.apache.nifi.processors.aws.s3.ListS3.restoreState(ListS3.java:133) ~[na:na]
	at org.apache.nifi.processors.aws.s3.ListS3.onTrigger(ListS3.java:155) ~[na:na]
	at org.apache.nifi.processor.AbstractProcessor.onTrigger(AbstractProcessor.java:27) ~[nifi-api-0.6.0-SNAPSHOT.jar:0.6.0-SNAPSHOT]
	at org.apache.nifi.controller.StandardProcessorNode.onTrigger(StandardProcessorNode.java:1139) [nifi-framework-core-0.6.0-SNAPSHOT.jar:0.6.0-SNAPSHOT]
	at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:139) [nifi-framework-core-0.6.0-SNAPSHOT.jar:0.6.0-SNAPSHOT]
	at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:49) [nifi-framework-core-0.6.0-SNAPSHOT.jar:0.6.0-SNAPSHOT]
	at org.apache.nifi.controller.scheduling.TimerDrivenSchedulingAgent$1.run(TimerDrivenSchedulingAgent.java:124) [nifi-framework-core-0.6.0-SNAPSHOT.jar:0.6.0-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471) [na:1.7.0_80]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:304) [na:1.7.0_80]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:178) [na:1.7.0_80]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [na:1.7.0_80]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) [na:1.7.0_80]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) [na:1.7.0_80]
	at java.lang.Thread.run(Thread.java:745) [na:1.7.0_80]

adamonduty commented:
@jskora Nice catch, I fixed the state issue and it seemed to work in my tests. I also rebased against master, added the recent proxy attributes, and changed the attribute `s3.storageClass` to `s3.storeClass` since that is the attribute used by `PutS3Object`.

joewitt commented:
Anyone in a good position to test/validate this?

jskora commented:
I can, but it might not be until later this week.

On Mon, Apr 18, 2016 at 11:26 PM, Joe Witt <notifications@github.com> wrote:

> Anyone in a good position to test/validate this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/nifi/pull/238#issuecomment-211711025>
>

apiri commented:
reviewing

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/ListS3.java:
```diff
@@ -0,0 +1,232 @@
+@TriggerWhenEmpty
+@InputRequirement(Requirement.INPUT_FORBIDDEN)
+@Tags({"Amazon", "S3", "AWS", "list"})
+@CapabilityDescription("Retrieves a listing of objects from an S3 bucket. For each object that is listed, creates a FlowFile that represents \"\n" +
```
There are some extraneous \s and +s that were folded into your description presumably from the IDE auto escaping.

adamonduty replied:
Fixed!

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/ListS3.java:
```diff
@@ -0,0 +1,232 @@
+                final Map<String, String> attributes = new HashMap<>();
+                attributes.put(CoreAttributes.FILENAME.key(), objectSummary.getKey());
+                attributes.put("s3.bucket", objectSummary.getBucketName());
+                attributes.put("s3.owner", objectSummary.getOwner().getId());
```
Seem to be getting NPEs fairly consistency regardless if the S3 bucket has files put into it from NiFi or those already existing in scenarios where a bucket may be publicly available.  Constraining permissions to an account using associated credentials seems to remedy this issue.
```
java.lang.NullPointerException: null
        at org.apache.nifi.processors.aws.s3.ListS3.onTrigger(ListS3.java:195) ~[nifi-aws-processors-0.6.0-SNAPSHOT.jar:0.6.0-SNAPSHOT]
        at org.apache.nifi.processor.AbstractProcessor.onTrigger(AbstractProcessor.java:27) ~[nifi-api-0.6.0-SNAPSHOT.jar:0.6.0-SNAPSHOT]
        at org.apache.nifi.controller.StandardProcessorNode.onTrigger(StandardProcessorNode.java:1139) [nifi-framework-core-0.6.0-SNAPSHOT.jar:0.6.0-SNAPSHOT]
        at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:139) [nifi-framework-core-0.6.0-SNAPSHOT.jar:0.6.0-SNAPSHOT]
        at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:1) [nifi-framework-core-0.6.0-SNAPSHOT.jar:0.6.0-SNAPSHOT]
        at org.apache.nifi.controller.scheduling.TimerDrivenSchedulingAgent$1.run(TimerDrivenSchedulingAgent.java:124) [nifi-framework-core-0.6.0-SNAPSHOT.jar:0.6.0-SNAPSHOT]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_77]
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) [na:1.8.0_77]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [na:1.8.0_77]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [na:1.8.0_77]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_77]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_77]
        at java.lang.Thread.run(Thread.java:745) [na:1.8.0_77]
```

adamonduty replied:
Good catch - looks like `getOwner()` can <a href="http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/model/S3ObjectSummary.html#getOwner()">return null</a>. Should be fixed now.

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/ListS3.java:
```diff
@@ -0,0 +1,232 @@
+            .build();
+
+    public static final List<PropertyDescriptor> properties = Collections.unmodifiableList(
+            Arrays.asList(BUCKET, KEY, REGION, ACCESS_KEY, SECRET_KEY, CREDENTIALS_FILE,
```
Doesn't seem like KEY is being used, if that is the case, should be removed from properties.

adamonduty replied:
Thanks, it wasn't needed.

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/ListS3.java:
```diff
@@ -0,0 +1,232 @@
+    private void persistState(final ProcessContext context) {
+        Map<String, String> state = new HashMap<>();
+        state.put(CURRENT_TIMESTAMP, String.valueOf(currentTimestamp));
+        int i = 0;
```
This seems like it might be a bit off as "i" never gets incremented and the state would just be overwriting for each of the keys currently in there.  What is the intended approach?

adamonduty replied:
Another good catch. I missed the increment, so it would always set `CURRENT_KEY_PREFIX+"0"`. Fixed now.

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/FetchS3Object.java:
```diff
@@ -46,7 +46,7 @@
 import com.amazonaws.services.s3.model.S3Object;
 
 @SupportsBatching
```
Should also be a @TriggerSerially given the state you have below and the inconsistencies that could be generated in multithreaded scenarios.

adamonduty replied:
I think maybe wires were crossed here, since this is a modification to `FetchS3Object` and not `ListS3`?

adamonduty replied:
That is, I intended to modify `FetchS3Object` in this way, and `ListS3` already uses `@TriggerSerially`.

apiri replied:
Wires crossed indeed! Thanks for clarifying.

apiri commented on nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/ListS3.java:
```diff
@@ -0,0 +1,232 @@
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) {
+        try {
+            restoreState(context);
```
Seems a bit heavy handed to have to restore state from the state manager on each trigger.  What are your thoughts on managing a map within the processor, which has its contents persisted after each execution and only performing a restore state in an @OnScheduled?

apiri commented:
Overall, seems to function well with some minor adjustments needed.  Would like to hash out some of the items with managing state.  Thanks for tackling this, we are certainly overdue for such functionality and I know many will be pleased to see this coming out in the next release.

adamonduty commented:
@apiri Thanks for the review! All your comments should be addressed except the state management pieces.

State management should work more or less like `ListHDFS`. I personally found the state management bits of `ListHDFS` difficult to follow, so I tried a similar but different approach.

On each restore, we grab two keys: `CURRENT_TIMESTAMP` and `CURRENT_KEY_PREFIX+"0"`. If both of these exist, the state is considered valid, and further restoration is attempted. Mostly that means we parse `CURRENT_TIMESTAMP` into a long and use `extractKeys()` to gather as many keys that start with `CURRENT_KEY_PREFIX` as are present in the state map. These values are later used to decided whether a listed key is newer than last time we ran. 

At the end of each run, we persist the newest key timestamp and at least one filename with the maximum discovered timestamp.

The processor does restore state on each trigger. I think there are a couple things to note:
* Listing an S3 bucket generally takes a little while (e.g. 100ms), so that puts a pretty high lower bound on the number of times state might be restored per second. In other words, it seems unlikely we'd try to restore state with a high frequency.
* ListS3 also persists state after each trigger. I didn't think about this until after submitting this PR, but really we should be committing flowfiles and persisting state after every batch (1000 objects by default) to avoid unbounded memory growth on large buckets. So in some cases the initial restore is only one of potentially many roundtrips to the state manager per trigger. 
* I have no concept of how expensive restoring and persisting state actually are.

I'd be happy to implement state management caching if you think its necessary. Its probably not complicated, but my general philosophy is that more conditionals = more bugs.

apiri commented:
@adamonduty Thanks for the explanation and your thoughts, very helpful and very much appreciated. I think things will be a little clearer now that the increment item is covered.  Going to do a build and some testing with the processor and will follow up, but certainly understand your points.

apiri commented:
@adamonduty Had some time to scope it out.  Things are looking great for the templates and items I laid out from before.  Looks like rebase went fine, so I shall work on getting this merged in shortly.  This will almost certainly be a new favorite processor 😄

apiri commented:
Merged to 0.x and master.  Fix Versions applied for 0.7 and 1.0

adamonduty commented:
Awesome, thanks again!! 👍
